tests/10apidoc/13ui-auth.pl

```diff
@@ -0,0 +1,271 @@
+         $user,
+         device_id => $DEVICE_ID,
+         initial_device_display_name => "device display",
+      )->then( sub {
+         # Initiate the interactive authentication session with the first device.
+         matrix_delete_device( $user, $DEVICE_ID, {} );
+      })->main::expect_http_401->then( sub {
+         my ($resp) = @_;
+
+         my $body = decode_json $resp->content;
+
+         log_if_fail("Response to empty body", $body);
+
+         assert_json_keys($body, qw(session params flows));
+         ref $body->{flows} eq "ARRAY" or die "Expected 'flows' as a list";
+
+         # Note that this uses the unstable value.
+         die "org.matrix.login.sso was not listed" unless
+            any { $_->{stages}[0] eq "org.matrix.login.sso" } @{ $body->{flows} };
+
+         Future->done( 1 );
+      });
+   };
+
+test "Can perform interactive authentication with SSO",
+   requires => [
+      local_user_fixture( with_events => 0 ),
+      $main::API_CLIENTS[0],
+      $main::HOMESERVER_INFO[0],
+   ],
+
+   do => sub {
+      my ( $user, $http, $homeserver_info ) = @_;
+
+      my $DEVICE_ID = "login_device";
+
+      my ($user_localpart) = $user->user_id =~ m/@([^:]*):/;
+      my $CAS_SUCCESS = <<"EOF";
+<cas:serviceResponse xmlns:cas='http://www.yale.edu/tp/cas'>
+    <cas:authenticationSuccess>
+         <cas:user>$user_localpart</cas:user>
+         <cas:attributes></cas:attributes>
+    </cas:authenticationSuccess>
+</cas:serviceResponse>
+EOF
+
+      # the ticket our mocked-up CAS server "generates"
+      my $CAS_TICKET = "goldenticket";
+      my $session;
+
+      # Create a device.
+      matrix_login_again_with_user(
+         $user,
+         device_id => $DEVICE_ID,
+         initial_device_display_name => "device display",
+      )->then( sub {
+         # Initiate the interactive authentication session via device deletion.
+         matrix_delete_device( $user, $DEVICE_ID, {} );
+      })->main::expect_http_401->then( sub {
+         my ( $resp ) = @_;
+
+         my $body = decode_json $resp->content;
+
+         log_if_fail( "Response to empty body", $body );
+
+         assert_json_keys( $body, qw( session params flows ));
+
+         $session = $body->{session};
+
+         # Note that we skip almost all of the CAS flow since it isn't important
+         # for this test. The user just needs to end up back at the homeserver
+         # with a valid ticket (and the original UI Auth session ID).
+         my $login_uri = $homeserver_info->client_location . "/_matrix/client/r0/auth/cas/ticket?session=$session&ticket=$CAS_TICKET";
+
+         Future->needs_all(
+            wait_for_cas_request(
+               "/cas/proxyValidate",
+               response => $CAS_SUCCESS,
+            ),
+            $http->do_request_json(
+               method   => "GET",
+               full_uri => $login_uri,
+               max_redirects => 0, # don't follow the redirect
+            ),
+         );
```

Comment:
since we've got two copies of this, it would be nice to factor it and the common parts of `CAS_SUCCESS` out to a separate function.

Reply:
That makes sense! I'll need to figure out how to do that, but shouldn't be too hard.